onnx/backend/test/case/node/resize.py

```diff
@@ -8,15 +8,233 @@
+        get_neighbor_idxes(4.6, 3, 10) == [4, 5, 6]
+        get_neighbor_idxes(4.4, 1, 10) == [4]
+        get_neighbor_idxes(4.6, 1, 10) == [5]
+        :param x:
+        :param n: the number of the wanted indexes
+        :param limit: the maximum value of index
+        :return: An np.array containing n nearest indexes in ascending order
+        """
+        idxes = sorted(range(limit), key=lambda idx: (abs(x - idx), idx))[:n]
+        idxes = sorted(idxes)
+        return np.array(idxes)
+
+    def get_neighbor(x, n, data):  # type: (float, int, np.ndarray) -> np.ndarray
+        """
+        Pad `data` in 'edge' mode, and get n nearest elements in the padded array and their indexes in the original
+        array
+        :param x:
+        :param n:  the number of the wanted elements
+        :param data: the array
+        :return: A tuple containing the indexes of neighbor elements (the index can be smaller than 0 or higher than
+        len(data)) and the value of these elements
+        """
+        pad_width = np.ceil(n / 2).astype(np.int)
+        padded = np.pad(data, pad_width, mode='edge')
+        x += pad_width
+
+        idxes = get_neighbor_idxes(x, n, len(padded))
+        ret = padded[idxes]
+        return idxes - pad_width, ret
+
+    input_width = len(data)
+    output_width = scale_factor * input_width
+    if scaler == 'align_corners':
+        if output_width == 1:
+            x_ori = 0.
+        else:
+            x_ori = x * (input_width - 1) / (output_width - 1)
+    elif scaler == 'asymmetric':
+        x_ori = x / scale_factor
+    elif scaler == 'tf_crop_and_resize':
+        if output_width == 1:
+            x_ori = (roi[1] - roi[0]) * (input_width - 1) / 2
+        else:
+            x_ori = x * (roi[1] - roi[0]) * \
+                (input_width - 1) / (output_width - 1)
+        x_ori += (roi[0] * (input_width - 1))
+        # Return extrapolation_value directly as what TF CropAndResize does
+        if x_ori < 0 or x_ori > input_width - 1:
+            return extrapolation_value
+    elif scaler == 'tf_half_pixel_for_nn':
+        x_ori = (x + 0.5) / scale_factor
+    elif scaler == 'pytorch_half_pixel':
+        if output_width == 1:
+            x_ori = -0.5
+        else:
+            x_ori = (x + 0.5) / scale_factor - 0.5
+    else:  # scaler == 'half_pixel'
+        x_ori = (x + 0.5) / scale_factor - 0.5
+    x_ori_int = np.floor(x_ori).astype(np.int).item()
+
+    # ratio must be in (0, 1] since we prefer the pixel on the left of `x_ori`
+    if x_ori.is_integer():
+        ratio = 1
+    else:
+        ratio = x_ori - x_ori_int
+
+    coeffs = get_coeffs(ratio)
+    n = len(coeffs)
+
+    idxes, points = get_neighbor(x_ori, n, data)
+
+    if exclude_outside:
+        for i, idx in enumerate(idxes):
+            if idx < 0 or idx >= input_width:
+                coeffs[i] = 0
+        coeffs /= sum(coeffs)
+
+    return np.dot(coeffs, points).item()
```

Comment:
Your code looks very inspiring. It basically decompose Resize into several abstractive steps:

- roi selection
- coordinate transformation
- neighbor selection
- dot product (interpolation)
It's not a blocking comment but I am wondering if we can further decompose Resize in the future.